### PR TITLE
Html console

### PR DESCRIPTION
**Pull request by Aloso**
This displays values logged with `console.log`, `console.warn`, `console.count`, etc. in the "output" panel on the right.

<table>
<tr><td>

**EDIT:**
I also replaced `style-loader` with `mini-css-extract-plugin`, so CSS is loaded faster.

</td></tr>
</table>

Example:

```js
console.log("hello world", { hello: "world" }, [Symbol("hello"), function world() {}])
console.log(window)
```

console

The console works similar to browser dev tools. All values are logged to the dev tools in addition to this console.

It currently doesn't support string interpolation or custom CSS, but I believe these features are rather niche.

### CODE REVIEW

File: src/main.ts [hunk 1]
```diff
@@ -129,19 +134,32 @@ async function compile() {
...
         } else {
-          document.getElementById("eval-output").textContent = "Main function not found. It is defined and public?";
+          logger.log("Main function not found. It is defined and public?");
         }
+        logger.unmountGlobally();
       });
     } else {
       erlangEditor.setValue(files.Ok["build/dev/erlang/gleam-wasm/main.erl"]);
 
       document.getElementById("eval-output").textContent = "Compiled successfully!\n\nNote that the Erlang target is not executable in the browser.";
```
Comment (JohnDoneth) [inline src/main.ts]: I believe this would need to use `logger.log` as well?

Comment (Aloso) [reply src/main.ts]: You are correct. I'll fix it in a moment.

**Comment by Aloso** [comment]
@JohnDoneth I believe this is ready to merge.

File: .vscode/settings.json [hunk 2]
```diff
@@ -0,0 +1,3 @@
+{
+    "liveServer.settings.root": "/dist"
```
Comment (JohnDoneth) [inline .vscode/settings.json]: I think we should ignore adding specific VSCode settings? Might want to ignore `.vscode` in `.gitignore`.

File: public/index.html [hunk 3]
```diff
@@ -1,51 +1,57 @@
...
-        <meta name="viewport" content="width=device-width, initial-scale=1.00, maximum-scale=1.00, minimum-scale=1.00, user-scalable=no"/>
-    </head>
-    <body>
-        <div class="spa">
-            <div class="header">
-                <h2>Gleam Playground</h2>
-                <a href="https://github.com/JohnDoneth/gleam-playground">
-                    <img src="github.png" class="gh-link"/>
-                </a>
+
```
Comment (JohnDoneth) [inline public/index.html]: As an aside: These formatting changes remind me that we should add this file to the prettier search path for CI.

File: src/index.css [hunk 4]
```diff
@@ -252,3 +252,128 @@ body,
 .output-panel.hidden {
   display: none;
 }
+
+.log-container {
+  font-family: monospace, monospace; /* stylelint-disable-line font-family-no-duplicate-names */
```
Comment (JohnDoneth) [inline src/index.css]: I believe this is equivalent to `font-family: monospace`.

File: src/index.css [hunk 5]
```diff
@@ -252,3 +252,128 @@ body,
...
+  font-size: 85%;
+  user-select: text;
+  margin: 0 10px;
+  border: 1px solid #aaa;
+  overflow: auto;
+  white-space: pre-wrap;
+  max-height: calc(60vh - 100px);
+}
+
+.log-container [data-log-level],
```
Comment (JohnDoneth) [inline src/index.css]: There's a lot of log specific styles here. Would you mind moving them into their own CSS file?

File: src/log.ts [hunk 6]
```diff
@@ -0,0 +1,690 @@
+/* eslint @typescript-eslint/ban-types: "off", @typescript-eslint/no-explicit-any: "off" */
+import { H, h, frag, styled } from "./virtualdom";
+
+export class HtmlLogger {
```
Comment (JohnDoneth) [inline src/log.ts]:
```suggestion
export class HTMLLogger {
```
I'd prefer this to be `HTMLLogger` to match with `HTMLElement` for consistency.

File: src/main.ts [hunk 7]
```diff
@@ -123,12 +124,14 @@ async function bundle(files) {
   return output[0].code;
 }
 
+const logger = new HtmlLogger(document.getElementById("eval-output"));
```
Comment (JohnDoneth) [inline src/main.ts]:
```suggestion
const logger = new HTMLLogger(document.getElementById("eval-output"));
```

File: src/main.ts [hunk 8]
```diff
@@ -4,6 +4,7 @@ import * as gleamWasm from "gleam-wasm";
 import { Notyf } from "notyf";
 import { registerGleam } from "./gleam";
 import * as monaco from "monaco-editor";
+import { HtmlLogger } from "./log";
```
Comment (JohnDoneth) [inline src/main.ts]:
```suggestion
import { HTMLLogger } from "./log";
```

File: public/index.html (hunk 3 above)
Comment (Aloso) [reply public/index.html]: I don't know how to do that, but I formatted the HTML file with prettier, so it should be fine now.

File: src/index.css (hunk 4 above)
Comment (Aloso) [reply src/index.css]: It is not. With `monospace`, both Firefox and Chrome use a smaller font size. With `monospace, monospace`, the font size is preserved and consistent across browsers.

File: src/index.css (hunk 5 above)
Comment (Aloso) [reply src/index.css]: Done.

File: src/log.ts (hunk 6 above)
Comment (Aloso) [reply src/log.ts]: Done.

File: .vscode/settings.json (hunk 2 above)
Comment (Aloso) [reply .vscode/settings.json]: I removed the file in the next commit.

**Comment by JohnDoneth** [comment]
Thanks for this :heart:

**Comment by Aloso** [comment]
You're welcome.

@JohnDoneth I was also thinking about making the output of Gleam types nicer and more informative. The output of lists is less than ideal ATM:

```rs
fn main() {
  [1, 2, 3, 4, 5]
}
```

Screenshot_20211222_202209

But I'm unsure how to show it instead. Should arrays and tuples look exactly as in Gleam? That would mean that JS arrays are displayed as `#(...)` and `List` objects are displayed as `[...]`.

**Comment by JohnDoneth** [comment]
Hmm. I can see the appeal of both choices. We could default to showing the "Gleam" types and have a toggle or something for displaying the raw JavaScript values.

**Pull request merged**